### PR TITLE
RavenDB-5226

### PR DESCRIPTION
**Pull request by karmeli87**

### CODE REVIEW

File: src/Raven.Client/Constants.cs [hunk 1]
```diff
@@ -210,6 +210,8 @@ public class Versioning
                 public const string EnableVersioning = "Raven-Enable-Versioning";
 
                 public const string DisableVersioning = "Raven-Disable-Versioning";
+
+                public const string RevisionedDocument = "Raven-Revisioned-Document";
```
Comment (ppekrol) [inline src/Raven.Client/Constants.cs]: Can't we use flags? https://github.com/ravendb/ravendb/blob/v4.0/src/Raven.Server/Documents/DocumentFlags.cs#L9

Comment (karmeli87) [reply src/Raven.Client/Constants.cs]: hm, it's indeed probably better to use flags here.

File: src/Raven.Server/Documents/Replication/ReplicationBatchDocumentItem.cs [hunk 2]
```diff
@@ -22,5 +22,48 @@ public class ReplicationBatchDocumentItem
         public DocumentFlags Flags;
         public short TransactionMarker;
         public long LastModifiedTicks;
+
```
Comment (ayende) [inline src/Raven.Server/Documents/Replication/ReplicationBatchDocumentItem.cs:25]: Don't use implicit operators, they are a good way to get into head scratching contests.
Use explicit `From` method overloads.

File: src/Raven.Server/Documents/Replication/ReplicationDocumentSender.cs [hunk 3]
```diff
@@ -30,112 +32,94 @@ public ReplicationDocumentSender(Stream stream, OutgoingReplicationHandler paren
...
+                if (current.MoveNext() == false)
+                {
+                    _workEnumerators.Remove(current);
+                }
+                return true;
+            }
+
+            public void Reset()
+            {
+                _workEnumerators.Clear();
```
Comment (ayende) [inline src/Raven.Server/Documents/Replication/ReplicationDocumentSender.cs]: Just throw `NotSupportedException` here

File: src/Raven.Server/Documents/Replication/ReplicationDocumentSender.cs [hunk 4]
```diff
@@ -30,112 +32,94 @@ public ReplicationDocumentSender(Stream stream, OutgoingReplicationHandler paren
...
+                _workEnumerators.AddRange(_initEnumerators);
+
+                foreach (var enumerator in _workEnumerators)
+                {
+                    enumerator.Reset();
+                    enumerator.MoveNext();
+                }
+            }
+
+            public void Clear()
```
Comment (ayende) [inline src/Raven.Server/Documents/Replication/ReplicationDocumentSender.cs]: Who will call `Clear`?

File: src/Raven.Server/Documents/Replication/ReplicationDocumentSender.cs [hunk 5]
```diff
@@ -30,112 +32,94 @@ public ReplicationDocumentSender(Stream stream, OutgoingReplicationHandler paren
...
-                var state = CurrentEnumerationState.None;
-
-                if (docsIt.MoveNext())
-                    state |= CurrentEnumerationState.HasDocs;
-                if (tombsIt.MoveNext())
-                    state |= CurrentEnumerationState.HasTombs;
-                if (conflictsIt.MoveNext())
-                    state |= CurrentEnumerationState.HasConflicts;
-                while (true)
+                _mergedInEnumerator.Clear();
```
Comment (ayende) [inline src/Raven.Server/Documents/Replication/ReplicationDocumentSender.cs]: Just create a new instance.

File: src/Raven.Server/Documents/Replication/ReplicationDocumentSender.cs [hunk 6]
```diff
@@ -30,112 +32,94 @@ public ReplicationDocumentSender(Stream stream, OutgoingReplicationHandler paren
             _parent = parent;
         }
 
-        [Flags]
-        private enum CurrentEnumerationState
+        public class MergedReplicationBatchEnumerator : IEnumerator<ReplicationBatchDocumentItem>
```
Comment (ayende) [inline src/Raven.Server/Documents/Replication/ReplicationDocumentSender.cs:35]: 👍

File: src/Raven.Server/Documents/DocumentsStorage.cs [hunk 7]
```diff
@@ -1985,6 +1988,17 @@ public void DeleteWithoutCreatingTombstone(DocumentsOperationContext context, st
                         ref oldValue, newEtag);
                 }
 
+                if (collectionName.IsSystem == false && (flags & DocumentFlags.Artificial) != DocumentFlags.Artificial)
+                {
+                    bool hasVersion =
+                        _documentDatabase.BundleLoader.VersioningStorage?.PutFromDocument(context, collectionName,
+                        key, document,changeVector) ?? false;
```
Comment (ayende) [inline src/Raven.Server/Documents/DocumentsStorage.cs]: Remove change vector from here, we'll have a new version

File: src/Raven.Server/Documents/Versioning/VersioningStorage.cs [hunk 8]
```diff
@@ -305,6 +346,45 @@ public IEnumerable<Document> GetRevisionsAfter(DocumentsOperationContext context
...
+            {
+                yield return TableValueToDocument(context, ref tvr.Reader);
+            }
+        }
+
+        private bool CheckIfVersionExists(DocumentsOperationContext context,string key, ChangeVectorEntry[] changeVector)
+        {
+            var table = context.Transaction.InnerTransaction.OpenTable(DocsSchema, RevisionDocuments);
+            Slice keySlice;
+            DocumentKeyWorker.GetSliceFromKey(context, key, out keySlice);
```
Comment (ayende) [inline src/Raven.Server/Documents/Versioning/VersioningStorage.cs]: This corrupts memory

File: src/Raven.Server/Documents/Versioning/VersioningStorage.cs [hunk 9]
```diff
@@ -305,6 +346,45 @@ public IEnumerable<Document> GetRevisionsAfter(DocumentsOperationContext context
...
+            {
+                Slice changeVectorSlice;
+                using (Slice.External(context.Allocator, (byte*)changeVectorPtr, sizeof(ChangeVectorEntry) * changeVector.Length,
+                    out changeVectorSlice))
+                {
+                    *(keySlice.Content.Ptr + keySlice.Size) = Seperator;
+                    changeVectorSlice.CopyTo(keySlice.Content.Ptr + keySlice.Size + 1);
+                }
+            }
+            foreach (var tvr in table.SeekForwardFrom(DocsSchema.Indexes[KeyAndEtagSlice],keySlice, startsWith: true))
```
Comment (ayende) [inline src/Raven.Server/Documents/Versioning/VersioningStorage.cs]: It's not starting with here.

**Review by ayende — CHANGES_REQUESTED** [review]

**Review by ayende — CHANGES_REQUESTED** [review]
please fix those issues

File: src/Raven.Server/Documents/DocumentsStorage.cs (hunk 7 above)
Comment (karmeli87) [reply src/Raven.Server/Documents/DocumentsStorage.cs]: If we remove this, we will send on replication the same document with different change vector.
Which will result on the destination as a document with 2 different versions.

**Pull request merged**